packages/actor-rdf-resolve-hypermedia-links-queue-priority/lib/LinkQueuePriority.ts

```diff
@@ -0,0 +1,229 @@
+import type { ILink, ILinkQueue } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+
+/**
+ * A link queue based on priority, using binary heap.
+ */
+export class LinkQueuePriority implements ILinkQueue {
+  /**
+   * Max heap with links and their priorities
+   */
+  public readonly links: ILink[] = [];
+  /**
+   * Data structure to track URLs in the queue and allow for constant time lookup of links in queue based
+   * on link URL
+   */
+  public readonly urlToLink: Record<string, ILink> = {};
+
+  /**
+   * Pushes element to heap by appending it to array and up-heaping the new element
+   */
+  public push(link: ILink): boolean {
+    const idx: number = this.links.length;
+
+    // If we push a link that has no metadata or has metadata but no priority we set priority to 0
+    // and always set index to end of array (and upheap from there)
+    if (!link.metadata || !link.metadata.priority) {
+      link.metadata = { priority: 0, index: idx };
+    } else {
+      link.metadata.index = idx;
+    }
+    this.links.push(link);
+
+    // Add to Records to allow fast updates in priority
+    this.urlToLink[link.url] = link;
+    this.upHeap(idx);
+    return true;
+  }
+
+  /**
+   * Pops the highest priority link and returns it. Then it sets the last element as root and
+   * down-heaps it.
+   * @returns popped element
+   */
+  public pop(): ILink {
+    const max = this.links[0];
+    const endArray = this.links.pop();
+
+    // If we remove link, remove it from records to reflect new state of queue
+    if (max) {
```

Comment:
Do we really need `max`? Can't we just check if `endArray` is truthy?

Reply:
I assume either would work, I chose max because we're operating on max.  I don't have a strong preference for either. I will add a check for the length of links, to make sure that empty pops don't do unnecessary checks